### DISCUSSION TITLE
Set/alter command working issues in customized model

### DISCUSSION THREAD

**Discussion by PandaRP01** [CURENT/andes#514] [Q&A]
HI Team

 I am developing a customized  model(i.e. TAPCON) to change the tap ratio of the `Line` model. 
To change the tap ration of the `Line` model, I defined as follows
method-1
```
for ii in range(self.n):
            self.system.Line.set(src='itap', idx=self.rfidx.v[ii], attr='v', value=self.irtapp.v[ii])
```
or

Method-2

```
for ii in range(self.n):
            self.system.Line.alter(src='itap',idx=self.rfidx.v[ii], value=self.irtapp.v[ii])
```

In both the way, it is unable to change the` itap` in the` Line` model.

Also, when I run `andes prepare -i`, it is generating the code properly. Also, during` tds` simulation, the output of the `TAPCON` model, i..e` irtapp` is changing as expected.

Could you please highlight my mistake in this?

Regards
Rakesh Panda

**Comment by cuihantao** [comment]
Hi Rakesh, where did you place these code in the model? The models are mostly descriptive, and those in `__init__()` will not be executed in run time. You might need to use other hook functions like `g_numeric()`

**Reply by PandaRP01** [reply]
Hello Dr. Cui

Thank you for your response. 

I have created a new folder inside models and placed my new customized model at thata folder. There I created __init__.py and added the model their.

Also, I added my model  in __init__. py (inside models) and also group.py .

Could you please explain .. how to use g_numeric() ??

I will try to explore this function from my side too. 


Regards
Rakesh panda

**Reply by cuihantao** [reply]
The folder and file organization do not matter for how the program is executed.

See here for an example. In `g_numeric`, you are updating the residuals by your custom function. The Jacobian will need to be updated accordingly. https://github.com/CURENT/andes/blob/3adca81feca5e986e52e578afabd33d7c56da7b3/andes/core/block.py#L736

If you are only looking to adjust the tap as the program runs based on some logic, you may use the perturbation file. See https://docs.andes.app/en/latest/getting_started/formats/disturbance.html

**Reply by PandaRP01** [reply]
Hello Dr. Cui

Thank you for this help. It helps me to write my own `g_numeric()` and` j_numeric()`.

However, when my `J_numeric()` function running, it is showing errors as follows:

```
TypeError                                 Traceback (most recent call last)
Cell In[10], line 7
      4 #ssPV.TDS.config.qrt = 1
      5 #ssPV.TDS.config.max_iter = 100
      6 ssPV.TDS.config.tf =50
----> 7 ssPV.TDS.run()

File ~\Andes_dev\andes\routines\tds.py:352, in TDS.run(self, no_summary, **kwargs)
    350 # only initializing at t=0 allows to continue when `run` is called again.
    351 if system.dae.t == 0:
--> 352     self.init()
    353 else:  # resume simulation
    354     self.init_resume()

File ~\Andes_dev\andes\routines\tds.py:209, in TDS.init(self)
    206 system.set_output_subidx(models=system.exist.pflow_tds)
    208 system.dae.clear_ts()
--> 209 system.store_sparse_pattern(models=system.exist.pflow_tds)
    210 system.store_adder_setter(models=system.exist.pflow_tds)
    211 system.store_no_check_init(models=system.exist.pflow_tds)

File ~\Andes_dev\andes\system.py:1161, in System.store_sparse_pattern(self, models)
   1159 for row, col, val in mdl.triplets.zip_ijv(jname):
   1160     print("in:",val,"row:", row,"col: ",col)
-> 1161     ii.extend(row)
   1162     jj.extend(col)
   1163     vv.extend(np.zeros_like(row))

TypeError: 'int' object is not iterable

```

How to overcome this error?
I have another question, could you please help me on "how existing ANDES model is updating `J-numer`ic and `g_numeric()` function automatically without writing `J-numeric()` inside the model file."? How the linkage has been established to do this operation?
